src/main/java/org/kainos/ea/team2/api/AuthenticationService.java

```diff
@@ -0,0 +1,171 @@
+            final String jwtSecret,
+            final IValidator<BasicCredentials> validator) {
+        if (jwtSecret == null) {
+            throw new IllegalArgumentException("JWT_SECRET NOT SET");
+        }
+
+        this.authSource = authSource;
+        this.jwtSigner = HMACSigner.newSHA256Signer(jwtSecret);
+        this.jwtVerifier = HMACVerifier.newVerifier(jwtSecret);
+        this.validator = validator;
+    }
+
+    /**
+     * Re-generates a password hash given the passed parameters.
+     * @param password the password to hash.
+     * @param salt the salt used in the hash.
+     * @param iterations the number of iterations used in the hash.
+     * @return byte[] representing the hash.
+     * @throws AuthenticationException throw if hashing fails
+     */
+    private byte[] getPasswordHash(
+            final String password,
+            final byte[] salt,
+            final int iterations)
+            throws AuthenticationException {
+
+      try {
+          final char[] passwordChars = password.toCharArray();
+          KeySpec spec =
+                  new PBEKeySpec(passwordChars, salt, iterations, KEY_LENGTH);
+          SecretKeyFactory factory = SecretKeyFactory.getInstance(HASH_ALG);
+
+          return factory.generateSecret(spec).getEncoded();
+      } catch (Exception e) {
+          throw new AuthenticationException(e.getMessage());
+      }
+    }
+
+    /**
+     * Attempts to generate a JWT with the credential provided.
+     * @param credentials the credentials to authenticate.
+     * @return JWT if successful, null if failed.
+     * @throws AuthenticationException Thrown on server error.
+     * @throws ValidationException Thrown on validation error
+     */
+    @Override
+    public JWT authenticate(
+            final BasicCredentials credentials)
+            throws AuthenticationException, ValidationException {
+
+        String error = validator.validate(credentials);
+        if (error != null) {
+            throw new ValidationException(error);
+        }
+
+        HashedPassword hashedPassword =
+                authSource.getHashedPasswordForUser(credentials.getUsername());
+
+        if (hashedPassword == null) {
+            return null;
+        }
+
+        //re hash the provided password and add the existing salt
+        String password = credentials.getPassword();
+        byte[] salt = hashedPassword.getSalt();
+        int iterations = hashedPassword.getIterations();
+        byte[] credentialHash = getPasswordHash(password, salt, iterations);
+
+        byte[] passwordHash = hashedPassword.getHashedPassword();
+
+        if (!Arrays.equals(credentialHash, passwordHash)) {
+            return null;
+        }
+
+        return new JWT()
+            .setIssuer("org.kainos.ea")
+            .setIssuedAt(ZonedDateTime.now(ZoneOffset.UTC))
+            .setSubject(credentials.getUsername())
+            .setExpiration(ZonedDateTime.now(ZoneOffset.UTC).plusDays(1));
```

Comment:
you could also include the role in the token so you wouldn't need to check the db